lib/Serialization/Deserialization.cpp

```diff
@@ -4098,7 +4106,11 @@ llvm::Error DeclDeserializer::deserializeDeclAttributes() {
         auto *diffAttr = DifferentiableAttr::create(
             ctx, isImplicit, SourceLoc(), SourceRange(), linear,
             /*parsedParameters*/ {}, jvp, vjp, /*trailingWhereClause*/ nullptr);
-        diffAttr->setParameterIndices(indices);
+        // Cache parameter indices so that they can set later.
+        // `DifferentiableAttr::setParameterIndices` cannot be called here
+        // because it requires `DifferentiableAttr::setOriginalDeclaration` to
+        // be called first.
+        diffAttrParamIndicesMap[diffAttr] = indices;
```

Comment:
If you remove the `Decl *decl` from the request, then you can do the `setParameterIndices` here, I think!